docs/source/basic/abstraction.rst

```diff
@@ -0,0 +1,275 @@
+On a GPU there are global, constant, shared and other memory types which all can be accessed explicitly.
+The interface has to abstract from these differences without sacrificing speed on any platform.
+
+A process running on a multi-socket node is the largest entity within *alpaka*.
+The abstraction is only about the task and data parallel execution on the process/node level and down.
+It does not provide any primitives for inter-node communication.
+However, such libraries can be combined with *alpaka*.
+
+An application process always has a main thread and is by definition running on the host.
+It can access the host memory and various accelerator devices.
+Such accelerators can be GPUs, Intel Xeon Phis, the host itself or other devices.
+Thus, the host not necessarily has to be different from the accelerator device used for the computations.
+For instance, an Intel Xeon Phi simultaneously can be the host and the accelerator device.
+
+The *alpaka* library can be used to offload the parallel execution of task and data parallel work simultaneously onto different accelerator devices.
+
+Task Parallelism
+----------------
+
+One of the basic building blocks of modern applications is task parallelism.
+For example, the operating system scheduler, deciding which thread of which process gets how many processing time on which CPU core, enables task parallelism of applications.
+It controls the execution of different tasks on different processing units.
+Such task parallelism can be, for instance, the output of the progress in parallel to a download.
+This can be implemented via two threads executing two different tasks.
+
+The valid dependencies between tasks within an application can be defined as a DAG (directed acyclic graph) in all cases.
+The tasks are represented by nodes and the dependencies by edges.
+In this model, a task is ready to be executed if the number of incoming edges is zero.
+After a task finished it's work, it is removed from the graph as well as all of it's outgoing edges,.
+This reduces the number of incoming edges of subsequent tasks.
+
+The problem with this model is the inherent overhead and the missing hardware and API support.
+When it is directly implemented as a graph, at least all depending tasks have to be updated and checked if they are ready to be executed after a task finished.
+Depending on the size of the graph and the number of edges this can be a huge overhead.
+
+*OpenCL* allows to define a task graph in a somewhat different way.
+Tasks can be enqueued into an out-of-order command queue combined with events that have to be finished before the newly enqueued task can be started.
+Tasks in the command queue with unmet dependencies are skipped and subsequent ones are executed.
+The ``CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE`` property of a command queue is an optional feature only supported by few vendors.
```

Comment:
We really need a source here. `CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE` has been present in the OpenCL specification since version 1.0. There is no mention of it being optional as far as I can see.

Reply:
Same here. A general overwork of the whole abstraction section is necessary to bring it up to date and make it more readable. But for this PR it was planned to make only some small changes to the text to speed up the merging.